test/e2e/kubectl.go

```diff
@@ -0,0 +1,215 @@
+		By("creating a replication controller")
+		runKubectl("create", "-f", nautilusPath)
+		validateController(updateDemoSelector, updateDemoContainer, nautilusImage, 2, 30*time.Second)
+		By("scaling down the replication controller")
+		runKubectl("resize", "rc", "update-demo-nautilus", "--replicas=1")
+		validateController(updateDemoSelector, updateDemoContainer, nautilusImage, 1, 30*time.Second)
+		By("scaling up the replication controller")
+		runKubectl("resize", "rc", "update-demo-nautilus", "--replicas=2")
+		validateController(updateDemoSelector, updateDemoContainer, nautilusImage, 2, 30*time.Second)
+	})
+
+	It(fmt.Sprint("should do a rolling update of a replication controller"), func() {
+		nautilusPath := filepath.Join(updateDemoRoot, "nautilus-rc.yaml")
+		kittenPath := filepath.Join(updateDemoRoot, "kitten-rc.yaml")
+
+		// Cleanup all resources incase we fail somewhere in the middle
+		defer cleanup(updateDemoRoot, updateDemoSelector)
+
+		By("creating the initial replication controller")
+		runKubectl("create", "-f", nautilusPath)
+		validateController(updateDemoSelector, updateDemoContainer, nautilusImage, 2, 30*time.Second)
+		By("rollingupdate to new replication controller")
+		runKubectl("rollingupdate", "update-demo-nautilus", "--update-period=1s", "-f", kittenPath)
+		validateController(updateDemoSelector, updateDemoContainer, kittenImage, 2, 30*time.Second)
+	})
+
+}
+
+func cleanup(filePath, selector string) {
+	By("using stop to clean up resources")
+	runKubectl("stop", "-f", filePath)
+
+	resources := runKubectl("get", "pods,rc", "-l", selector, "--no-headers").stdout.String()
+	if resources != "" {
+		Fail(fmt.Sprintf("Resources left running after stop:\n%s", resources))
+	}
+}
+
+func validateController(selector, container, image string, replicas int, timeout time.Duration) {
+
+	getPodsTemplate := "--template={{range.items}}{{.id}} {{end}}"
+
+	// NB: kubectl adds the "exists" function to the standard template functions.
+	// This lets us check to see if the "running" entry exists for each of the containers
+	// we care about. Exists will never return an error and it's safe to check a chain of
+	// things, any one of which may not exist. In the below template, all of info,
+	// containername, and running might be nil, so the normal index function isn't very
+	// helpful.
+	// This template is unit-tested in kubec{tl|fg}, so if you change it, update the unit test.
+	//
+	// You can read about the syntax here: http://golang.org/pkg/text/template/
+	getContainerStateTemplate := fmt.Sprintf(`--template={{and (exists . "currentState" "info" "%s" "state" "running")}}`, container)
+
+	getImageTemplate := fmt.Sprintf(`--template={{(index .currentState.info "%s").image}}`, container)
+
+	getHostIPTemplate := "--template={{.currentState.hostIP}}"
+
+	By(fmt.Sprintf("waiting for all containers in %s pods to come up.", selector))
+	for start := time.Now(); time.Since(start) < timeout; time.Sleep(5 * time.Second) {
```

Comment:
The other pattern I saw was a naive counter, but that doesn't include time spent in calls. Some of these calls are a bit longer running (e.g. rollingupdate), so this seemed better.